alephium/lib/deployment.ts

```diff
@@ -0,0 +1,389 @@
+    const codeHash = cryptojs.SHA256(bytecode).toString()
+    const previous = deployContractResults.get(contract.typeId)
+    const tokens = params.initialTokenAmounts ? getTokenRecord(params.initialTokenAmounts) : undefined
+    const issueTokenAmount: string | undefined = params.issueTokenAmount?.toString()
+    const needToDeploy = await needToDeployContract(
+      signer.provider,
+      previous,
+      params.initialAttoAlphAmount,
+      tokens,
+      issueTokenAmount,
+      codeHash
+    )
+    if (needToDeploy) {
+      const result = await contract.transactionForDeployment(signer, params)
+      await signer.submitTransaction(result.unsignedTx, result.txId)
+      const confirmed = await waitTxConfirmed(signer.provider, result.txId)
+      const deployContractResult: DeployContractResult = {
+        fromGroup: result.fromGroup,
+        toGroup: result.toGroup,
+        txId: result.txId,
+        blockHash: confirmed.blockHash,
+        contractId: result.contractId,
+        contractAddress: result.contractAddress,
+        codeHash: codeHash,
+        attoAlphAmount: params.initialAttoAlphAmount,
+        tokens: tokens,
+        issueTokenAmount: issueTokenAmount
+      }
+      deployContractResults.set(contract.typeId, deployContractResult)
+      return deployContractResult
+    }
+    // we have checked in `needToDeployContract`
+    return previous!
+  }
+
+  const runScript = async (script: Script, params: RunScriptParams, taskName?: string): Promise<RunScriptResult> => {
+    const bytecode = script.buildByteCodeToDeploy(params.initialFields ? params.initialFields : {})
+    const codeHash = cryptojs.SHA256(bytecode).toString()
+    const key = taskName ? taskName : script.typeId
+    const previous = runScriptResults.get(key)
+    const tokens = params.tokens ? getTokenRecord(params.tokens) : undefined
+    const needToRun = await needToRunScript(signer.provider, previous, params.attoAlphAmount?.toString(), tokens, codeHash)
+    if (needToRun) {
+      const result = await script.transactionForDeployment(signer, params)
+      await signer.submitTransaction(result.unsignedTx, result.txId)
+      const confirmed = await waitTxConfirmed(signer.provider, result.txId)
+      const runScriptResult: RunScriptResult = {
+        fromGroup: result.fromGroup,
+        toGroup: result.toGroup,
+        txId: result.txId,
+        blockHash: confirmed.blockHash,
+        codeHash: codeHash,
+        attoAlphAmount: params.attoAlphAmount?.toString(),
+        tokens: tokens
+      }
+      runScriptResults.set(key, runScriptResult)
+      return runScriptResult
+    }
+    // we have checked in `needToRunScript`
+    return previous!
+  }
+
+  const getDeployContractResult = (typeId: string): DeployContractResult => {
+    const result = deployContractResults.get(typeId)
+    if (result === undefined) {
+      throw new Error(`Contract ${typeId} deployment result does not exist`)
```

Comment:
[Nit]: Deployment result of contract "${typeId}" does not exist

Reply:
Thanks!